### DISCUSSION TITLE
More granular imports

### DISCUSSION THREAD

**Discussion by bhashemian** [Project-MONAI/MONAI#4274] [Ideas]
Some codes in MONAI import many components from the parent modules (look at example below). This is generally fine but can increase the chance of circular import in MONAI core. Although we should avoid such situations with some tricks like deferred imports, it would be a good practice to import the modules with more granularity in the core MONAI.

This is an example of what we can see in MONAI:
```python
from monai.utils import InterpolateMode, deprecated_arg, ensure_tuple, ensure_tuple_rep, min_version, optional_import
from monai.utils.enums import PostFix
```
and a better way to import would be:
```python
from monai.utils.deprecate_utils import deprecated_arg
from monai.utils.enums import InterpolateMode, PostFix
from monai.utils.misc import ensure_tuple, ensure_tuple_rep
from monai.utils.module import min_version, optional_import
```

The changes are small and easy per each file but there are more than 200 instances of this kind of imports, which makes it a tedious task to improve. I suggest to look for this imports in any PR review and gradually fix the previous ones.

**Comment by bhashemian** [comment]
@wyli @Nic-Ma @ericspod, what do you think?

**Comment by ericspod** [comment]
I'm not sure this will fix circular import issues, if you import `foo.bar.baz` Python will import `foo.bar` first. Regardless we should be doing it with low level imports or high level and not mixing them going forward. For users in notebooks it's better example to show the low level ones so people know where things are found but more convenient when typing to do it high level perhaps.

**Comment by wyli** [comment]
so this is mostly style preference, unless we have automatic tools to enforce this, it's not easy to implement. I'm converting it to a discussion.

**Comment by ericspod** [comment]
I'm not aware of tools that will refactor imports for us like that, isort just does sorting. I can see it being implementing using refactoring tools like [rope](https://github.com/python-rope/rope) but that would be a script we'd have to write.
